app/models/ssh_execution_provider.rb

```diff
@@ -18,27 +18,51 @@ def supports_effective_user?
       true
     end
 
+    def find_ip_or_hostname(host)
+      interface = effective_interfaces(host).find { |i| i.ip.present? }
+      if interface
+        interface.ip
+      else
+        host.fqdn
+      end
+    end
+
     private
 
     def ssh_user(host)
       host.params['remote_execution_ssh_user']
     end
 
     def ssh_port(host)
-      Integer(host.params['remote_execution_ssh_port'] || Setting[:remote_execution_ssh_port])
+      Integer(host_setting(host, :remote_execution_ssh_port))
     end
 
     def effective_user(template_invocation)
       template_invocation.effective_user
     end
 
     def effective_user_method(host)
-      method = host.params['remote_execution_effective_user_method'] || Setting[:remote_execution_effective_user_method]
+      method = host_setting(host, :remote_execution_effective_user_method)
       unless EFFECTIVE_USER_METHODS.include?(method)
         raise _('Effective user method "%{current_value}" is not one of %{valid_methods}') %
                   { :current_value => method, :valid_methods => EFFECTIVE_USER_METHODS}
       end
       method
     end
+
+    private
+
+    def effective_interfaces(host)
+      return [] unless host_setting(host, :remote_execution_connect_by_ip)
```

Comment:
Good point. Please have a look at the latest version. I hope it should have everything right now.